mozilla_django_oidc/contrib/drf.py

```diff
@@ -0,0 +1,110 @@
+"""
+Classes/functions for integrating with Django REST Framework.
+
+http://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
+"""
+
+from django.contrib.auth import get_backends
+from django.core.exceptions import ImproperlyConfigured, SuspiciousOperation
+from rest_framework import authentication, exceptions
+from requests.exceptions import HTTPError
+
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+from mozilla_django_oidc.utils import parse_www_authenticate_header
+
+
+def guess_oidc_backend():
+    """
+    Look through AUTH_BACKENDS to find one that is or extends the mozilla
+    OIDC backend class.
+    """
+    backends = [
+        backend for backend in get_backends()
+        if isinstance(backend, OIDCAuthenticationBackend)
+    ]
+    if not backends:
+        raise ImproperlyConfigured('No backends extending OIDCAuthenticationBackend found!')
+    if len(backends) > 1:
+        raise ImproperlyConfigured('More than one OIDCAuthenticationBackend found!')
+    return backends[0]
+
+
+class OIDCAuthentication(authentication.BaseAuthentication):
+    """
+    Provide OpenID authentication for DRF.
+    """
+
+    # used by the authenticate_header method.
+    www_authenticate_realm = 'api'
+
+    def __init__(self, backend=None):
+        self.backend = backend or guess_oidc_backend()
+
+    def authenticate(self, request):
+        """
+        Authenticate the request and return a tuple of (user, token) or None
+        if there was no authentication attempt.
+        """
+        access_token = self.get_access_token(request)
+
+        if not access_token:
+            return None
+
+        try:
+            user = self.backend.get_or_create_user(access_token, None, None)
+        except HTTPError as exc:
+            resp = exc.response
+
+            # if the oidc provider returns 401, it means the token is invalid.
+            # in that case, we want to return the upstream error message (which
+            # we can get from the www-authentication header) in the response.
+            if resp.status_code == 401 and 'www-authenticate' in resp.headers:
+                data = parse_www_authenticate_header(resp.headers['www-authenticate'])
+                raise exceptions.AuthenticationFailed(data['error_description'])
+
+            # for all other http errors, just re-raise the exception.
+            raise
+        except SuspiciousOperation as exc:
+            # trust that the message in the exception is presentable to the user
+            raise exceptions.AuthenticationFailed('Login failed: %s' % exc)
```

Comment:
I would prefer to avoid sending the actual exception to the end user response. Let's add a generic message there or just leave only the error code.

Reply:
There is no error code, but I'll change it to just a generic "Login failed" and log the actual exception (without a stacktrace) instead.